exercises/protein-translation/protein_translation_test.go

```diff
@@ -102,71 +89,115 @@ func TestCodon(t *testing.T) {
 		}
 		t.Logf("PASS: Protein translation test: %s", test.input)
 	}
+
+	for _, test := range codonInvalidBaseCases {
+		actual, err := FromCodon(test.input)
+		if !errors.As(err, &ErrInvalidBase{}) {
+			t.Fatalf("FAIL: Protein translation test: %s\nExpected type error: %T\nActual type error: %T",
+				test.input, ErrInvalidBase{}, err)
+		}
+		if actual != test.expected {
+			t.Fatalf("FAIL: Protein translation test: %s\nExpected: %s\nActual: %s",
+				test.input, test.expected, actual)
+		}
+		t.Logf("PASS: Protein translation test: %s", test.input)
+	}
+
+	for _, test := range codonStopCases {
+		actual, err := FromCodon(test.input)
+		if !errors.As(err, &ErrStop{}) {
+			t.Fatalf("FAIL: Protein translation test: %s\nExpected type error: %T\nActual type error: %T",
+				test.input, ErrStop{}, err)
+		}
+		if actual != test.expected {
+			t.Fatalf("FAIL: Protein translation test: %s\nExpected: %s\nActual: %s",
+				test.input, test.expected, actual)
+		}
+		t.Logf("PASS: Protein translation test: %s", test.input)
+	}
 }
 
 type rnaCase struct {
-	input         string
-	expected      []string
-	errorExpected error
+	input    string
+	expected []string
 }
 
-var proteinTestCases = []rnaCase{
+var proteinSuccessCases = []rnaCase{
 	{
 		"AUGUUUUCUUAAAUG",
 		[]string{"Methionine", "Phenylalanine", "Serine"},
-		nil,
 	},
 	{
 		"AUGUUUUGG",
 		[]string{"Methionine", "Phenylalanine", "Tryptophan"},
-		nil,
 	},
 	{
 		"AUGUUUUAA",
 		[]string{"Methionine", "Phenylalanine"},
-		nil,
 	},
 	{
 		"UGGUGUUAUUAAUGGUUU",
 		[]string{"Tryptophan", "Cysteine", "Tyrosine"},
-		nil,
 	},
+}
+
+var proteinInvalidBaseCases = []rnaCase{
 	{
 		"UGGAGAAUUAAUGGUUU",
 		[]string{"Tryptophan"},
-		ErrInvalidBase,
 	},
 }
 
 func TestProtein(t *testing.T) {
-	for _, test := range proteinTestCases {
+	for _, test := range proteinSuccessCases {
 		actual, err := FromRNA(test.input)
-		if test.errorExpected != nil {
-			if test.errorExpected != err {
-				t.Fatalf("FAIL: RNA translation test: %s\nExpected error: %q\nActual error: %q",
-					test.input, test.errorExpected, err)
-			}
-		} else if err != nil {
-			t.Fatalf("FAIL: RNA translation test: %s\nExpected: %s\nGot error: %q",
+		if err != nil {
+			t.Fatalf("FAIL: Protein translation test: %s\nExpected: %s\nGot error: %q",
 				test.input, test.expected, err)
 		}
 		if !reflect.DeepEqual(actual, test.expected) {
 			t.Fatalf("FAIL: RNA Translation test: %s\nExpected: %q\nActual %q", test.input, test.expected, actual)
 		}
 		t.Logf("PASS: RNA translation test: %s", test.input)
 	}
+	for _, test := range proteinInvalidBaseCases {
+		actual, err := FromRNA(test.input)
+		if !errors.As(err, &ErrInvalidBase{}) {
+			t.Fatalf("FAIL: Protein translation test: %s\nExpected type error: %T\nActual type error: %T",
+				test.input, ErrInvalidBase{}, err)
+		}
+		if !reflect.DeepEqual(actual, test.expected) {
```

Comment:
If the function doesn't return the expected error for an invalid input, that's the end of the test, isn't it? No need to go on and compare the data value against expectations: there are none.

Reply:
My idea here was to ensure that when an error occurs, the function doesn't return anything besides the error.

Reply:
Well, that's a good idea in principle, but in practice, a non-nil error says that the data value should be ignored, and that's what we should do in the test, too.

Reply:
That makes sense, we should expect that the code does this.